qa/loopback/src/paqa.c

```diff
@@ -909,56 +909,54 @@ static int PaQa_SingleLoopBackTest( UserOptions *userOptions, TestParameters *te
 }
 
 /*******************************************************************/
-static void PaQa_SetDefaultTestParameters( TestParameters *testParamsPtr, PaDeviceIndex inputDevice, PaDeviceIndex outputDevice )
+static TestParameters PaQa_GetTestParameters( PaDeviceIndex inputDevice, PaDeviceIndex outputDevice, const UserOptions* userOptions )
```

Comment:
- `struct TestParameters` is rather large. I do not agree with returning it by-value. Neither does the internet:
https://stackoverflow.com/questions/30980759/when-should-i-pass-or-return-a-struct-by-value

- The verb "Make" is better than "Get" here.

So my suggestion and request is to go with:

```
static void PaQa_MakeTestParameters( TestParameters *testParamsPtr, PaDeviceIndex inputDevice, PaDeviceIndex outputDevice, const UserOptions* userOptions )
```

Reply:
> `struct TestParameters` is rather large. I do not agree with returning it by-value. Neither does the internet:       https://stackoverflow.com/questions/30980759/when-should-i-pass-or-return-a-struct-by-value

The only *potential* advantage of passing that struct as a pointer argument is performance. This is very obviously not a performance-sensitive code path so that argument doesn't apply here. (Even if it did apply, it would not make any difference in this case, because when returning a larger struct, the compiler passes it to the function as hidden pointer behind the scenes, so it will all compile to the same thing anyway.)

There is, however, a significant *readability* argument in favor of returning the struct: it makes it clear what the inputs and outputs of the function are. When passing pointers, it's not necessarily obvious at the callsite which pointees are used as input parameters and which ones are used as output parameters.

Returning a struct also makes it possible to use the return value to initialize a `const` variable, to make it clear that the variable is not changed after initialization. This is another readability advantage.

More arguments in favour of preferring return values to output parameters are laid out in [Abseil TotW #176](https://abseil.io/tips/176) and [C++ Core Guidelines F.20](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out) (both of which are for C++ but most of the points apply to C too).

> The verb "Make" is better than "Get" here.

Agreed. Done.

Reply:
PortAudio is a C89 project. C++ guidelines have no bearing here.

Reply:
> PortAudio is a C89 project. C++ guidelines have no bearing here.

The arguments I brought forward in my response, as well as most of the points raised in the links I posted (e.g. literally all the bullet points in the introduction of [Abseil TotW #176](https://abseil.io/tips/176)) apply to C just as well as they apply to C++. Just because it says "C++" on the title doesn't mean the arguments presented lose all merit. The reason why I used C++ sources is because I usually work in C++ so I already had these references at hand.

I am disappointed that you chose to dismiss the points I raised without even attempting to engage in constructive discussion. So far you have not presented specific, concrete reasons as to why this approach is problematic in this particular case.

It is demotivating for contributors to get pushback of their PRs for seemingly arbitrary reasons, and for them to be forced to downgrade the quality of their patches for no other reason than "the maintainer told me to". I don't think this is conducive to attracting contributors to PortAudio in the long term.